2016/exercises/final_project/Alba-Petar-Rebecca-Simon/PolytopeProject/Gale_Diagram.cpp

```diff
@@ -0,0 +1,322 @@
+int Gale_Diagram::orientation(const Point &x, const Point &a, const Point &b) const{
+    Matrix M = Matrix(vector<Point>{b.minus(a),x.minus(a)});
+    return sgn(M.determinant());
+}
+//Checks if onePoint lies in the triangle defined by threePoints
+bool Gale_Diagram::in_triangle(const vector<Point> &onePoint, const vector<Point> &triangle) const{
+    Point x=onePoint[0];
+    Point a=triangle[0];
+    Point b=triangle[1];
+    Point c=triangle[2];
+    int o1 = orientation(x,a,b);
+    int o2 = orientation(x,b,c);
+    int o3 = orientation(x, c,a);
+    return o1==o2 && o2 == o3;
+}
+// returns whether both linesegments (vectors of size 2) intersect.
+bool Gale_Diagram::intersectingLineSegments(const vector<Point> &lineSegment1, const vector<Point> &lineSegment2) const{
+    Point a=lineSegment1[0];
+    Point b=lineSegment1[1];
+    Point c=lineSegment2[0];
+    Point d=lineSegment2[1];
+    int o1 = orientation(a, c, d);
+    int o2 = orientation(b,c,d);
+    int o3 = orientation(c,a,b);
+    int o4 = orientation(d,a,b);
+    return o1!=o2 && o3!=o4;
+}
+bool Gale_Diagram::check_intersecting(const vector<Point> &points) const{
+    if(points.size()!=4) throw runtime_error("Wrong Size in check_intersecting, Gale_diagram");
+    
+    vector<Point> positives;
+    vector<Point> negatives;
+    for(Point p : points){
+        if(p.get_sign()) positives.push_back(p);
+        else negatives.push_back(p);
+    }
+    int s = (int) positives.size();
+    if(s==0 || s==4) return false;
+    if(s==1) return in_triangle(positives, negatives);
+    if(s==2) return intersectingLineSegments(positives, negatives);
+    return in_triangle(negatives, positives);
+    return true;
+}
+
+void Gale_Diagram::test() const{
+    Point a(vector<double>{0,0},true);
+    Point b(vector<double>{0,1},true);
+    Point c(vector<double>{1,0},true);
+    Point d(vector<double>{0,-1},true);
+    Point e(vector<double>{-1,0},true);
+    Point f(vector<double>{1,1},true);
+    
+    cout << (orientation(b,a,c)==1) << (orientation(d,a,c)==-1) << (orientation(c,a,b)==-1);
+    cout << (orientation(e,a,b)==1) << endl;
+    
+    vector<Point> zero{a};
+    vector<Point> triangle{b,f,c};
+    vector<Point> triangle2{f,d,e};
+
+    cout << "triangle results " << (in_triangle(zero, triangle)==false) << (in_triangle(zero, triangle2)==true) << endl;
+    
+}
+
+
+bool Gale_Diagram::isSimplex(int i, int j , int k, int l) const{
+    //Remove the columns i,j,k,l by selecting the other columns:
+    vector<int> pointIndices = findMissing(i,j,k,l);
+    /*cout << i << " " << j << " " << k << " " << l << endl;
+    for(int m=0; m<pointIndices.size(); m++){
+		cout << pointIndices[m] << " ";
+	}
+	cout << endl;*/
+    //Select points
+    vector<Point> fourPoints;
+    for(int m=0; m<pointIndices.size(); m++){
+        int pointIndex = pointIndices[m];
+        Point p(points[pointIndex].get_coordinates(), points[pointIndex].get_sign());
```

Comment:
adapt this to the implementation of `findMissing`. In general, you don't need the detour via indices, you can work directly with the entries of the vector themselves